plutus-contract-exe/src/Language/Plutus/Contract.hs

```diff
@@ -0,0 +1,106 @@
+{-# LANGUAGE DeriveAnyClass     #-}
+{-# LANGUAGE DeriveGeneric      #-}
+{-# LANGUAGE DerivingStrategies #-}
+{-
+    Event-based interface between contract executables and the app platform and
+    (by extension) the wallet.
+
+    Two types of events are defined:
+
+    1. 'ContractOut'. Events produced by the contract for consumption by app
+       platform and wallet. Includes transactions and instructions to start
+       watching interesting addresses on the ledger.
+
+    2. 'LedgerUpdate'. Events that inform the contract about changes to the
+       ledger state.
+
+    Contracts will offer an HTTP interface with the following routes.
+
+    * 1 route per contract endpoint (as defined by contract author)
+    * 1 route for submitting 'LedgerUpdate' events
+
+    NOTE: With this design, there are two classes of input events, only one of
+          which has a proper data type: Input events from the ledger (in the
+          form of 'LedgerUpdate') and input events from the user (in the form
+          of HTTP endpoints).
+
+          In my opinion it would be cleaner to add a
+
+          'ContractEndpoint EndpointArgs'
+
+          constructor to 'LedgerUpdate' (and rename the type) that captures all
+          contract endpoints activated by the user. I think this will make
+          testing a lot easier. But we need to think about the 'EndpointArgs'
+          type and how it maps to the actual contract endpoints.
```

Comment:
We might need to give people some TH to generate the endpoints, but we likely want that anyway? I agree that there should be a datatype that represents a call to a contract endpoint with one constructor for each endpoint.

Reply:
Agree, I don't think we can do it without TH for generating endpoints, and that's what the plutus-contract-exe library will be about.

I'll bring up the other question (one HTTP endpoint per contract endpoint, or one constructor per contract endpoint) in the call tomorrow.

Reply:
Yeah, I don't know whether it's nicer to have one HTTP endpoint per contract endpoint. If we're sufficiently generic and the schemas are expressive enough then a HTTP endpoint like `runCommand :: Command -> State -> (Output, State)` might be sufficient, it's just that all of those will have to be big sum types.

Reply:
From the perspective of the app platform either option works.

Each contract endpoint will be a separate GraphQL endpoint for expressive type checking on method parameters. Whether each endpoint maps to a corresponding HTTP endpoint for the contract (as in the example), or if they are all called through a single endpoint, like defined by @michaelpj above, that would work for us as well.

Reply:
Mapping a HTTP endpoint to a contract endpoint seems like the most direct approach. No need to introduce an additional layer IMO